parley/src/layout/editor.rs

```diff
@@ -160,46 +168,59 @@ where
 
     /// Delete the selection or up to the next word boundary (typical ‘ctrl + delete’ behavior).
     pub fn delete_word(&mut self) {
-        let start = self.editor.selection.focus().text_range().start;
         if self.editor.selection.is_collapsed() {
-            let end = self
-                .editor
-                .cursor_at(start)
-                .next_word(&self.editor.layout)
-                .index();
-
-            self.editor.buffer.replace_range(start..end, "");
-            self.update_layout();
-            self.editor
-                .set_selection(self.editor.cursor_at(start).into());
+            let focus = self.editor.selection.focus();
+            let start = focus.index();
+            let end = focus.next_logical_word(&self.editor.layout).index();
+            if self.editor.text().get(start..end).is_some() {
+                self.editor.buffer.replace_range(start..end, "");
+                self.update_layout();
+                self.editor.set_selection(
+                    Cursor::from_byte_index(&self.editor.layout, start, Affinity::Downstream)
+                        .into(),
+                );
+            }
         } else {
             self.delete_selection();
         }
     }
 
     /// Delete the selection or the previous cluster (typical ‘backspace’ behavior).
     pub fn backdelete(&mut self) {
-        let end = self.editor.selection.focus().text_range().start;
         if self.editor.selection.is_collapsed() {
-            if let Some(start) = self
+            // Upstream cluster
+            if let Some(cluster) = self
                 .editor
                 .selection
                 .focus()
-                .cluster_path()
-                .cluster(&self.editor.layout)
-                .map(|x| {
-                    if self.editor.selection.focus().affinity() == Affinity::Upstream {
-                        Some(x)
-                    } else {
-                        x.previous_logical()
-                    }
-                })
-                .and_then(|c| c.map(|x| x.text_range().start))
+                .logical_clusters(&self.editor.layout)[0]
+                .clone()
             {
+                let range = cluster.text_range();
+                let end = range.end;
+                let start = if cluster.info().whitespace() == Whitespace::Newline
+                    || cluster.info().is_emoji()
```

Comment:
`info` is outside of the visibility boundary. That is, I believe the editor component is something we want to support customising externally (obvious example, for rich text editing, which is clearly out-of-scope). Therefore, it shouldn't have a dependency on anything `pub(crate)` in Parley.

One half of this condition is already public; I'd suggest adding a similar helper for the other half:
```suggestion
                let start = if cluster.is_hard_line_break()
                    || cluster.is_emoji()
```

Reply:
Added public `is_emoji()` and changed these conditions.